composer/models/huggingface.py

```diff
@@ -272,20 +273,30 @@ def loss(self, outputs, batch):
             return outputs[0]
 
     def eval_forward(self, batch, outputs: Optional[Any] = None):
-        output = outputs if outputs else self.forward(batch)
-        if self.use_logits:
+        if 'mode' in batch and batch['mode'] == 'lm_task':
             self.labels = batch.pop('labels')
-            if self.config.use_return_dict:
-                output = output['logits']
-            else:
-                # logits are at index 1 in the output tuple
-                output = output[1]
+            args = {'input_ids': batch['input_ids']}
+            args['attention_mask'] = ~(batch['input_ids'] == batch['eos_tok_id'])
```

Comment:
I think it would be safer if `attention_mask` was part of the batch, to match the same spec that HF default data collators produce. So basically, this line would move to `in_context_learning_evaluation.py`, and inside here, we wouldn't need to construct `args`, and just call `self.forward(batch)` on a HF-standard batch.

Reply:
The batch contains a bunch of keys that wouldn't be used by forward but are necessary for computing the metric, including the continuation indices, so I will need to do something special before calling forward either way.

One potential alternative is to group the input ids and attention mask into their own value within the batch, maybe under the key 'forward_input'?

Reply:
I think it makes sense if batches of type `mode='lm_task'` have a unique format, but it still feels like that format should be enforced by the dataset+collate_fn. Would different models need to create different attention_masks? I'm basically just trying to move this line: `args['attention_mask'] = ~(batch['input_ids'] == batch['eos_tok_id'])` so it doesn't have to be written in every model's `eval_forward()`.

Understood about using a subset of the batch keys, I think that's what we do in our [HF example](https://github.com/mosaicml/examples/blob/fd087710737ee56d0efed7671c6b7238961c0c0d/llm/src/hf_causal_lm.py#L66-L68) and it seems fine.